### PR TITLE
Gas enforcement policy requiring payment to meet estimated costs

### PR DESCRIPTION
**Pull request by tkporter**
* A bit of a refactor of the existing gas enforcement policy logic - now, there's still a single `GasPolicyEnforcer` that's shared by all message submitters. The GasPolicyEnforcer can be configured with a GasPaymentPolicy -- e.g. `GasPaymentPolicyNone` (consider all messages as having paid enough, i.e. always return true), `GasPaymentPolicyMinimum` (require messages to have paid at least some fixed amount, this is helpful for transitioning toward full gas payment requirements...), and `GasPaymentPolicyMeetsEstimatedCost`, which will fetch exchange rates and gas prices and ensure enough has been paid to cover the approximate cost of the transaction.
* The aforementioned `GasPaymentPolicyMeetsEstimatedCost` is new. It has logic very similar to the SDK's InterchainGasCalculator - it'll look at the USD price of the payment token (the origin chain native token), the USD price of the destination chain's token (the one that tx fees are actually paid in), and the destination chain's gas price, and calculate the approximate amount of origin chain native tokens to cover the destination chain tx costs.
* Note that we don't use any Gelato APIs for the cost estimation - this is after discovering their API only supports quoting fees in tokens on the *same* chain in which the tx is executed on, and unfortunately we require payment to be made in a different chain's native token than the one that the tx is executed on. Apparently this feature is added to their roadmap. This also lets us use the gas enforcement policy regardless of Gelato support.
* I also made a small refactor of the serial submitter so it's more clear when a message has been processed. I noticed a bug where if the gas enforcement policy rejected a message, it wasn't actually getting pushed back into the run queue. The refactor fixes this and makes it much more clear. I also did a quick drive by to fix https://github.com/abacus-network/abacus-monorepo/issues/573, where our existing logic will just consider the tx being mined as indicating that the message has been processed, but in reality the tx could revert. Now we will not consider a message as processed if the tx reverts.
* I've yet to write tests for`CoinGeckoCachingPriceGetter`

### CODE REVIEW

File: rust/abacus-base/src/settings/chains.rs
Comment (tkporter) [inline rust/abacus-base/src/settings/chains.rs:298]: I added these to add some compile-time checks on ensuring we always have a coingecko API ID for each mainnet chain. I'm curious how you feel about this vs the existing macro here: https://github.com/abacus-network/abacus-monorepo/blob/main/rust/abacus-core/src/chain.rs#L78

If you like this approach I'll move over the macro linked above so we only have one source of truth for abacus chains / domains, but otherwise I'm happy to get rid of this. Another idea is instead of having separate enums for mainnet / testnet / local test chains, is to just have one big enum with them all, and then have a runtime check for getting coingecko API IDs.

Just generally would love feedback here because I could see this not being very ergonomic

File: rust/abacus-core/src/chain.rs
Comment (tkporter) [inline rust/abacus-core/src/chain.rs:57]: I added these to add some compile-time checks on ensuring we always have a coingecko API ID for each mainnet chain. I'm curious how you feel about this vs the old macro

If this feels really weird to you I'm happy to get rid of this. Another idea is instead of having separate enums for mainnet / testnet / local test chains, is to just have one big enum with them all, and then have a runtime check for getting coingecko API IDs

File: rust/gelato/src/types.rs
Comment (mattiecnvr) [inline rust/gelato/src/types.rs:9]: You should just add an import.

**Review by mattiecnvr — APPROVED** [review]
The only thing I feel much conviction on is the enums being merged into one rather than 3, otherwise take from this what you will but approving so you don't have to try and get another review after I am out.

File: rust/agents/relayer/src/msg/gelato_submitter/sponsored_call_op.rs
Comment (tkporter) [reply rust/agents/relayer/src/msg/gelato_submitter/sponsored_call_op.rs:96]: 🤯

**Pull request merged**
